### PR TITLE
add sparseGPT pruner feature, refactor pruning class

### PR DESCRIPTION
**Pull request by WeiweiZhang1**
## Type of Change

feature
API not changed

## Description

refactor sparseGPT and retrain-free, add two new post-training pruning class.

### CODE REVIEW

**Review by n1ck-guo — APPROVED** [review]

File: neural_compressor/compression/pruner/pruners/sparse_gpt.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        
+        # self.criterion = get_criterion(config=self.config, modules=self.modules)
+        gc.collect()
+        self.gpts = {}
+        logger.warning("sparse_gpt pruner fixed the weights, Please DO NOT train or update gradients.")
+        assert "1x1" in self.pattern.pattern or ":" in self.pattern.pattern, \
+                "sparse_gpt pruner type only supports 1x1 and N:M patterns."
+    
+    class SparseGPT():
+        def __init__(self, module):
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners/sparse_gpt.py:46]: add reference paper later

File: neural_compressor/compression/pruner/pruning.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+    return register
+
+
+class BasePruning:
+    """Pruning.
+
+    The main class to do pruning; it contains at least one Pruner object.
+
+    Args:
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruning.py:52]: merge experimental/pruning.py later

File: neural_compressor/compression/pruner/pruning.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        self._model = model
+        self.pruners_info = config
+        self.pruners = self._generate_pruners(model)
+
+    def _generate_pruners(self, model: torch.nn.Module):
+        """Generate pruners.
+
+        :param config: WeightPruningConfig
+        :param model: The torch module to be pruned
+        :return: A list of pruner
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruning.py:74]: need to support inc prepare_compression later, I don't see  any change for this part

File: neural_compressor/compression/pruner/pruning.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+        :param config: WeightPruningConfig
+        :param model: The user's model, a torch.nn.Module object
+        :param opt: The user's optimizer, a torch.optim object
+        :return: The modified model and optimizer
+        """
+        self.dev = torch.device(type='cpu')
+        self.model_dev = self._model.device
+        self._layers = get_layers(self._model)
+        if torch.cuda.is_available():
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruning.py:258]: get the device from model

File: neural_compressor/compression/pruner/pruning.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    #     delattr(opt, "pruners")
+    #     opt.step = opt.orig_step
+    #     delattr(opt, "orig_step")
+    #     return model, opt
+
+
+@register_pruning("sparse_gpt_pruning")
+class SparseGPTPruning(BasePruning):
+    def __init__(self, config, model: torch.nn.Module,
+                 dataloader: torch.utils.data.DataLoader,
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruning.py:235]: remove the type from dataloader, users could customize their own

File: neural_compressor/compression/pruner/__init__.py [hunk 6]
```diff
@@ -19,7 +19,62 @@
...
+from typing import Optional, Union
 
+FRAMEWORK = {
+    'pytorch': 'pt',
+    'keras': 'keras'
+}
 
+def prepare_pruning(config, model: torch.nn.Module,
+                    opt_or_dataloader: Union[torch.optim.Optimizer, torch.utils.data.DataLoader],
+                    loss_func=None, framework='pytorch', device: str ='cuda:0'):
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/__init__.py:33]: delete the device from the function

File: neural_compressor/compression/pruner/__init__.py [hunk 7]
```diff
@@ -19,7 +19,62 @@
...
+from .utils import process_config, torch, logger
+from typing import Optional, Union
 
+FRAMEWORK = {
+    'pytorch': 'pt',
+    'keras': 'keras'
+}
 
+def prepare_pruning(config, model: torch.nn.Module,
+                    opt_or_dataloader: Union[torch.optim.Optimizer, torch.utils.data.DataLoader],
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/__init__.py:32]: if the model is limited to torch.nn.module, why do we need a framework parameter? better detect the framework automatically

**Review by wenhuach21 — CHANGES_REQUESTED** [review]
The current code is not suitable to support prepare_comperssion.

**Review by wenhuach21 — APPROVED** [review]

**Comment by chensuyue** [comment]
We can merge this PR first, please submit another PR to update docs including 1. example readme，2. example list, 3. pruning feature docs.

**Pull request merged**